types/rlp_unmarshal.go

```diff
@@ -312,75 +312,120 @@ func (t *Transaction) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 
 	p.Hash(t.Hash[:0], v)
 
+	// Setup defaults
+	t.GasPrice = new(big.Int)
+	t.Value = new(big.Int)
+	t.V = new(big.Int)
+	t.R = new(big.Int)
+	t.S = new(big.Int)
+	t.From = ZeroAddress
+
 	// nonce
 	if t.Nonce, err = elems[0].GetUint64(); err != nil {
 		return err
 	}
+
 	// gasPrice
-	t.GasPrice = new(big.Int)
-	if err := elems[1].GetBigInt(t.GasPrice); err != nil {
+	if err = elems[1].GetBigInt(t.GasPrice); err != nil {
 		return err
 	}
+
 	// gas
 	if t.Gas, err = elems[2].GetUint64(); err != nil {
 		return err
 	}
+
 	// to
 	if vv, _ := v.Get(3).Bytes(); len(vv) == 20 {
 		// address
 		addr := BytesToAddress(vv)
 		t.To = &addr
-	} else {
-		// reset To
-		t.To = nil
 	}
+
 	// value
-	t.Value = new(big.Int)
-	if err := elems[4].GetBigInt(t.Value); err != nil {
+	if err = elems[4].GetBigInt(t.Value); err != nil {
 		return err
 	}
+
 	// input
 	if t.Input, err = elems[5].GetBytes(t.Input[:0]); err != nil {
 		return err
 	}
 
 	// V
-	t.V = new(big.Int)
 	if err = elems[6].GetBigInt(t.V); err != nil {
 		return err
 	}
 
 	// R
-	t.R = new(big.Int)
 	if err = elems[7].GetBigInt(t.R); err != nil {
 		return err
 	}
+
 	// S
-	t.S = new(big.Int)
 	if err = elems[8].GetBigInt(t.S); err != nil {
 		return err
 	}
 
-	if len(elems) >= 10 {
+	if len(elems) > 9 {
 		// Type
-		if t.Type, err = ReadRlpTxType(elems[9]); err != nil {
+		if t.Type, err = readRlpTxType(elems[9]); err != nil {
 			return err
 		}
 
-		if t.IsStateTx() {
+		switch t.Type {
+		case StateTx:
+			if len(elems) != 11 {
+				return fmt.Errorf("incorrect number of elements to decode state tx, expected 11 but found %d", len(elems))
+			}
+
 			// We need to set From field for state transaction,
 			// because we are using unique, predefined address, for sending such transactions
 			// From
 			if vv, err := v.Get(10).Bytes(); err == nil && len(vv) == AddressLength {
 				// address
 				addr := BytesToAddress(vv)
 				t.From = addr
-			} else {
-				// reset From
-				t.From = ZeroAddress
+			}
+		case DynamicFeeTx:
+			if len(elems) != 12 {
+				return fmt.Errorf("incorrect number of elements to decode dynamic fee tx, expected 12 but found %d", len(elems))
+			}
+
+			t.GasFeeCap = new(big.Int)
+			t.GasTipCap = new(big.Int)
+
+			// gasFeeCap
+			if err = elems[10].GetBigInt(t.GasFeeCap); err != nil {
+				return err
+			}
+
+			// gasTipCap
+			if err = elems[11].GetBigInt(t.GasTipCap); err != nil {
```

Comment:
You didn't want to simplify like you did it for other `*big.Int` fields?

```suggestion
			// gasFeeCap
			if err = elems[10].GetBigInt(t.GasFeeCap); err != nil {
				return err
			}

			// gasTipCap
			if err = elems[11].GetBigInt(t.GasTipCap); err != nil {
```

Reply:
Done: https://github.com/0xPolygon/polygon-edge/pull/1111/commits/e4bba66b1a9a52de5025843186abd58460d72146